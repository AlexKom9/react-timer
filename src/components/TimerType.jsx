import React, { Component } from "react";
import * as constants from "../constants/constant";

class TimerType extends Component {
  constructor(props) {
    super(props);
    this.state = {
      inputField: props.types.map(elem=> elem.trim().toLowerCase()).join(", "),
      textAreaField: props.types.map(elem=> elem.trim().toLowerCase()).join("\n"),
      selectField: props.types.map(elem=>elem.trim().toLowerCase()),
    };
    this.refSelectField = React.createRef();
  }

  static defaultProps = { types: [constants.HOURS, constants.MINUTES, constants.SECONDS] };

  selectHandler = () => {
    const options = Array.prototype.slice.call(
      this.refSelectField.current.querySelectorAll("option")
    );
    const arr = options
      .filter(option => {
        if (option.selected) {
          return true;
        }
      })
      .map(option => option.value);

    this.setState({
      inputField: arr.join(", "),
      textAreaField: arr.join("\n"),
      selectField: arr
    }, this.setTimerType);
  };

  inputHandler = event => {
    const value = event.target.value;
    const arr = value.split(",").map(elem => elem.trim());
    this.setState({
      inputField: value,
      textAreaField: arr.map(elem => elem.trim().toLowerCase()).join("\n"),
      selectField: arr
    },
      this.setTimerType
    );
  };

  textAreaHandler = event => {
    const value = event.target.value;
    const arr = value.split("\n");
    this.setState({
      inputField: arr.map(elem => elem.trim().toLowerCase()).join(", "),
      textAreaField: value,
      selectField: arr
    },
      this.setTimerType
    );
  };

  setTimerType = () => {
    const { selectField } = this.state;
    let hours = false;
    let minutes = false;
    let seconds = false;
    if (selectField.map(elem => elem.toUpperCase()).includes(constants.HOURS)) {
      hours = true;
    }
    if (selectField.map(elem => elem.toUpperCase()).includes(constants.MINUTES)) {
      minutes = true;
    }
    if (selectField.map(elem => elem.toUpperCase()).includes(constants.SECONDS)) {
      seconds = true;
    }
    const timerType = {
      hours,
      minutes,
      seconds
    };
    console.log(timerType)
    this.props.setTimerType(timerType);
  };

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.state.inputField}
          onChange={this.inputHandler}
        />
        <br />
        <textarea
          value={this.state.textAreaField}
          onChange={this.textAreaHandler}
          cols="30"
          rows="10"
        />
        <br />
        <select
          multiple
          onChange={this.selectHandler}
          ref={this.refSelectField}
          value={this.state.selectField}
        >
          {this.props.types.map(type => (
            <option
              key={type}
              value={constants[type].toLowerCase()}
              selected={this.state.selectField.includes(constants[type])}
            >
              {constants[type].toLowerCase()}
            </option>
          ))}
        </select>
      </div>
    );
  }
}

TimerType.propTypes = {};

export default TimerType;
