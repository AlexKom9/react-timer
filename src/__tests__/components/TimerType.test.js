import React from "react";
import TimerType from "../../components/TimerType";
import { shallow, mount } from "enzyme";

describe(`TimerType`, () => {

  test(`initial reander`, () => {
    const wrapper = shallow(<TimerType setTimerType={() => {}} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("change input", () => {
    const value = "hours\nminutes\nseconds";
    const wrapper = mount(<TimerType setTimerType={()=> {}} />);
    const name = 'username';

    wrapper.find("input").simulate("change", {
      target: {
        name,
        value: "hours, minutes, seconds"
      }
    });

    expect(wrapper.find("textarea").props().value).toBe(value);
    // expect(wrapper.find("textarea").props().value).toBe(value);
    // debugger;
    console.log('option = ', wrapper.find("option[value='hours']").props());

  });
});


