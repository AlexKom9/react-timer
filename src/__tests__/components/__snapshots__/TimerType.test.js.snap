// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerType initial reander 1`] = `
"<div class=\\"timer-type\\"><input type=\\"text\\" class=\\"timer-type__input\\" value=\\"hours, minutes, seconds\\"/><br/><textarea class=\\"timer-type__text\\" cols=\\"30\\" rows=\\"10\\">hours
minutes
seconds</textarea><br/><select multiple=\\"\\" class=\\"timer-type__select\\"><option selected=\\"\\" value=\\"hours\\">hours</option><option selected=\\"\\" value=\\"minutes\\">minutes</option><option selected=\\"\\" value=\\"seconds\\">seconds</option></select></div>"
`;
